R/duplicates.R

```diff
@@ -0,0 +1,181 @@
+  checkmate::assert_data_table(dt)
+  checkmate::assert_character(col_names, null.ok = TRUE)
+  
+  if (is.null(col_names)) {
+    anyDuplicated(dt) != 0
+  } else {
+    checkmate::assert_subset(col_names, colnames(dt))
+    anyDuplicated(dt, by = col_names) != 0
+  }
+  
+}
+
+#' get columns in the assay data required to have unique data
+#' 
+#' get columns in the assay data required to have unique (non-duplicated) data
+#' 
+#' @param dt data.table with assay data
+#' @examples
+#' sdata <- get_synthetic_data("finalMAE_small")
+#' smetrics_data <- convert_se_assay_to_dt(sdata[[1]], "Metrics")
+#' get_assay_req_uniq_cols(smetrics_data)
+#' @return charvec with columns required to have unique data
+#' @keywords duplicates
+#' @export
+#' 
+get_assay_req_uniq_cols <- function(dt) {
+  
+  checkmate::assert_data_table(dt)
+  col_ids <- get_settings_from_json(
+    "assay_dt_req_uniq_col_ids",
+    system.file(package = "gDRutils", "settings.json")
+  )
+  
+  # check with both pretiffied and unprettified version of ids
+  col_names_p <- unlist(get_prettified_identifiers(col_ids, simplify = FALSE))
+  col_names_up <- as.character(get_env_identifiers(col_ids, simplify = FALSE))
+  col_names <- unique(c(col_names_p, col_names_up))
+  
+  intersect(col_names, names(dt))
+}
+
+#' check if assay data contains duplicated data
+#' 
+#' An auxiliary function that checks for duplicates in the assay data
+#' 
+#' @param dt data.table with assay data
+#' 
+#' @return logical flag indicating if a dt contains duplicated rows or not
+#' @keywords duplicates
+#' @examples
+#' sdata <- get_synthetic_data("finalMAE_small")
+#' smetrics_data <- convert_se_assay_to_dt(sdata[[1]], "Metrics")
+#' has_assay_dt_duplicated_rows(smetrics_data)
+#' @export
+#' 
+has_assay_dt_duplicated_rows <- function(dt) { 
+
+  checkmate::assert_data_table(dt)
+  
+  col_names <- get_assay_req_uniq_cols(dt)
+  has_dt_duplicated_rows(dt, col_names)
+
+}
+
+
+#' Helper function to find duplicated rows
+#'
+#' @param x DataFrame or data.table
+#' @param col_names character vector, columns in which duplication are searched for
+#' @param output string with the output format to be returned - one of "index" (index of duplicates) or "data" (subset of input data with duplicates)
+#' @examples
+#' dt <- data.table::data.table(a = c(1, 2, 3), b = c(3, 2, 2))
+#' get_duplicated_rows(dt, "b")
+#' get_duplicated_rows(dt, "b", output = "data")
+#' @return integer vector or data.table with duplicated rows
+#' @keywords duplicates
+#' @export
+get_duplicated_rows <- function(x, 
+                                col_names = NULL, 
+                                output = "index") {
+  
+  checkmate::assertMultiClass(x, c("data.table", "DataFrame"))
+  checkmate::assert_true(all(col_names %in% colnames(x)))
+  checkmate::assert_choice(output, c("index", "data"))
+  
+  
+  if (!is.null(col_names)) {
+    sub_x <- subset(x, select = col_names)
+  }
```

Comment:
```suggestion
  } else {
  sub_x <- x
 }
```

@gladkia Here is one omission.
```
  DF1co <- S4Vectors::DataFrame("Gnumber" = c("G0123456.1-1", "G0123456.2-2", "G1234567.1-1"),
                                "DrugName" = c("drug_name1", "drug_name1", "drug_name2"),
                                "Gnumber_2" = c("G9876543.1-1", "G9876543.1-1", "G9876543.1-1"),
                                "DrugName_2" = c("codrug_name1", "codrug_name1", "codrug_name1"),
                                "Concentration_2" = c("untreated", "untreated", "untreated"))

> get_duplicated_rows(DF1co)
Error in h(simpleError(msg, call)) : 
  error in evaluating the argument 'x' in selecting a method for function 'duplicated': object 'sub_x' not found
Called from: h(simpleError(msg, call))
```